CKEditor Blog Post (#40)
mechmotum/mechmotum.github.io: content/jupyter-summer-2020.rst

> Developing a Textbook Editor Plugin Inserting Executable Code Blocks

Reviewer: ```suggestion
Developing a LibreTexts Editor Plugin For Inserting Executable Code Blocks
```

====================================================================

> :date: 2020-06-13 00:00:00 

Reviewer: Change  date to date of publishing when ready to merge.

:tags: oer, education, jupyter, textbooks, engineering, libretexts 
:category: education 

> :slug: jupyter-summer-2020

Reviewer: ```suggestion
:slug: libretexts-jupyter-plugin

```

:authors: Hao Huang, Tannavee Kumar, Celine Liang 
:summary: Blog post on creating a CKEditor plugin which could insert 

>   executable code blocks

Reviewer: Needs to be indented to the B in Blog.

Background
----------

During the end of the Fall Quarter 2019 and Winter Quarter 2020, we focused
on building a `CKEditor Binder Plugin
<https://github.com/LibreTexts/ckeditor-binder-plugin>`__ to be used on

> the `LibreTexts <https://libretexts.org>`__ to allow textbook authors to

Reviewer: ```suggestion
the `LibreTexts <https://libretexts.org>`__ website to allow textbook authors to

```

insert executable code blocks. We intend for both textbook authors and
readers to have the ability to edit and run code blocks efficiently, making
online educational content more interactive.
 
Build Process
-------------
We started the project by surveying

> LibreTexts authors and readers on their most requested features. 

Reviewer: I think we shared this publicly, you could link to it here.

Author: added this link: https://docs.google.com/document/d/1eV08l_4djKJ7bc8r0LPbD5bp3QT7mHTZgABUleH15H0/edit

BinderHub
^^^^^^^^^
On the backend, the CKEditor Binder plugin utilizes a project called
`BinderHub  <https://binderhub.readthedocs.io/en/latest/>`__ to run code
blocks. BinderHub is developed as part of the Jupyter project and gives
custom computing environments based on a list of requirements specified
through a GitHub repository. We intend to use a BinderHub that we deployed
on our bare-metal Kubernetes cluster as our backend environment to run

> codeblocks.

Reviewer: Remove last sentence.

Thebelab
^^^^^^^^
`Thebelab <https://github.com/minrk/thebelab>`__ and  `Juniper
<https://github.com/ines/juniper>`__ are two examples of projects which can

> insert code blocks into HTML pages and running them by requesting a kernel

Reviewer: ```suggestion
insert Binder backed code blocks into HTML pages and running them by requesting a kernel

```

from a computing backend like BinderHub. We deliberated on which project to
incorporate into our plugin, and decided to use Thebelab since it seemed to

> be more actively maintained and had more community support.

Reviewer: I though there were some features we preferred.

Creating the plugin
^^^^^^^^^^^^^^^^^^^
Our plugin is based on the `CKEditor 4
<https://ckeditor.com/docs/ckeditor4/latest/>`__, an open source “what you
see is what you get” text editor. This editor is used on the LibreTexts

> website through a service called Mindtouch.

Reviewer: I'd just say, "this is the editor authors use on the LibreTexts website"

Our approach to this plugin is to make use of a `widget
<https://ckeditor.com/docs/ckeditor4/latest/guide/widget_sdk_intro.html>`__,
on the editor which allows us to place all the HTML elements of Thebelab

> together as one unit. Additionally, we created a `dialog window

Reviewer: Define widget in this context.

<https://ckeditor.com/docs/ckeditor4/latest/guide/dev_howtos_dialog_windows.html>`__
for each code block so that users can modify each block whenever they want.

Mindtouch Specific Settings
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> One of the challenges we faced was working around Mindtouch, which

Reviewer: Define what Mindtouch is and why it is important to discuss.

sometimes caused the plugin to function in unexpected ways. For instance,
Mindtouch seemed to apply its own CSS to the plugin. This caused text
overflow, addition of characters to the end of each line, etc. In order to
resolve this, we added our own `styling
<https://github.com/LibreTexts/ckeditor-binder-plugin/tree/staging/src/styles>`__
to the plugin. If one wishes to use the CKEditor plugin on their own pages,
they can remove the extra styling in the folder.

Another difference was javascript conflict. When we were trying to support
Jupyter Widgets, we found that it depended on Require.js. However, adding
Require.js to Mindtouch would break all the JQuery plugins. In order to fix
all the javascript conflicts, we created registerPlugin.js to include all
Mindtouch specific js code.

Including different programming languages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
During development, we used `environments developed by the Jupyter project
<https://github.com/binder-examples>`__. Using their Binder environments

> helped us test and include different languages in our editor. Ideally, we

Reviewer: Eventually we will utilize ...

available via ``CodeMirror`` for all the languages except ``SageMath``
which is not supported by ``CodeMirror``. 

It is important to note that we suggest authors ``run`` the sample ``Hello
World`` print statement first to get the kernel started, once that is
successful, code can be added in the dialog box. For ``C++``, if the code
has already been ``run``, but the author wants to make any changes, they
will need to ``restart`` the kernel in order to avoid an ``Interpreter
Error`` as any variables will be assigned more than once in ``Binder``
which is not allowed in ``C++``.

Packages and libraries can be exported as they normally would; however, if
an author finds that a specific package or library that they would like to
use is not currently available they can make a request to have it added by
either sending an `email to the Jupyter Team <jupyterteam@ucdavis.edu>`__,
linked in the dialog box, or `open an issue 
<https://github.com/LibreTexts/ckeditor-binder-plugin/issues>`__ . 

Once the author is ready to insert the code block into their textbook page,
they have the option to either ‘Insert with code and output,’ ‘Insert with

> code only,’ or ‘Insert with output only.’ Selection of any of these choices

Reviewer: Single ticks in rst do not do render in any special way. Did you want double backticks for literal?

Author: The single quotes are intentional, as we were quoting what shows up on the dialog box.

Reviewer: Those are backticks though. If you want single quotes use `'`.

depends on how the author intends to communicate the information provided
in the code block. If the code block has already been inserted into the
page, and the author wants to make any changes, they can simply double
click on that section and the dialog box will pop back up. 

.. figure:: https://objects-us-east-1.dream.io/mechmotum/example-libretexts-jupyter-page.gif
   :width: 100%
   :alt: GIF depicting an example page made using CKEditor
   :align: center

   *Figure 1: Example of what and author can do with CKEditor, and what the output will look like*

Future
------ 

As mentioned before, a long term goal is to use ``default-env`` for the

> packages, this is the same environment that is used for JupyterHub, and by

Reviewer: ```suggestion
packages, this is the same environment that is used for the LIbreTexts-UCD JupyterHub, and by

```

Developing a LibreTexts Editor Plugin For Inserting Executable Code Blocks

> ====================================================================

Reviewer: ```suggestion
==========================================================================

```

Developing a LibreTexts Editor Plugin For Inserting Executable Code Blocks
==========================================================================

:date: 2020-06-22 00:00:00 
:tags: oer, education, jupyter, textbooks, engineering, libretexts 
:category: education 
:slug: libretexts-jupyter-plugin

> 

Reviewer: I'm not sure this empty line is allowed.
